spring-core/src/main/java/org/springframework/core/io/buffer/DefaultDataBuffer.java

```diff
@@ -317,6 +323,43 @@ private void write(ByteBuffer source) {
 		this.writePosition += length;
 	}
 
+	@Override
+	public DataBuffer write(CharSequence charSequence, Charset charset) {
+		Assert.notNull(charSequence, "'charSequence' must not be null");
+		Assert.notNull(charset, "'charset' must not be null");
+		CharsetEncoder charsetEncoder = charset.newEncoder()
+				.onMalformedInput(CodingErrorAction.REPLACE)
+				.onUnmappableCharacter(CodingErrorAction.REPLACE);
+		CharBuffer inBuffer = CharBuffer.wrap(charSequence);
+		int estimatedSize = (int) (inBuffer.remaining() * charsetEncoder.averageBytesPerChar());
+		ensureCapacity(estimatedSize);
+		charsetEncoder.reset();
+		ByteBuffer outBuffer = asByteBuffer(this.writePosition, writableByteCount());
+		try {
+			for (; ; ) {
```

Comment:
I would take a look at `io.netty.buffer.ByteBufUtil#encodeString0`, and see what Netty does there. Their algorithm looks somewhat simpler.